### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by dominikvagner**

### CODE REVIEW

File: src/pages/Landing/index.tsx [hunk 1]
```diff
@@ -117,7 +118,7 @@ export default () => {
             onChange={(value) => setNewUser({ ...newUser, isCool: value })}
             isChecked={newUser.isCool}
           />
-          <Button type='submit'>Submit</Button>
+          <SnazzyButton isSnazzy={true} type='submit'>Submit</SnazzyButton>
```
Comment (Andrewgdewar) [inline src/pages/Landing/index.tsx]: #nit
isSnazzy={true} 
Could be: 
isSnazzy

[The default value of unspecified props is true](https://legacy.reactjs.org/docs/jsx-in-depth.html#props-default-to-true)

File: src/pages/Landing/index.tsx [hunk 2]
```diff
@@ -44,8 +44,28 @@ export default () => {
...
+    onSettled: () => queryClient.invalidateQueries({ queryKey: ['customers'] })
+  })
+
   const onSubmit = (e: FormEvent<Element>) => {
     e.preventDefault();
+    newUserMutation.mutate({
+      name: 'John Doe',
+      color: 'red',
+      age: 0,
+      isCool: false,
```
Comment (Andrewgdewar) [inline src/pages/Landing/index.tsx]: Defining defaults should be done at the component level, if so desired. 
The issue with doing this here is you are spreading the "newUser" below these values.
Meaning any value that is NOT undefined (including "" or null) will overwrite these.

**Comment by Andrewgdewar** [comment]
For solution 3:  
There is no form validation for the submit button.
When completing a form field for a modal or really any action that requires inputs from the user, we generally have some logic to prevent the submit button from being used until we have everything we need from them.

Although not specifically asked in the problem, I would normally expect to see this on any real ticket.

File: src/pages/Landing/index.tsx [hunk 3]
```diff
@@ -44,8 +44,28 @@ export default () => {
...
+    onMutate: async newCustomer => {
+      await queryClient.cancelQueries('customers');
+      const previousCustomers = queryClient.getQueryData('customers');
+      queryClient.setQueryData('customers', old => [...old as Customer[], newCustomer]);
+      return { previousCustomers };
+    },
+    onError: (err, newCustomer, context) => {
+      queryClient.setQueryData('todos', context?.previousCustomers)
+    },
+    onSettled: () => queryClient.invalidateQueries({ queryKey: ['customers'] })
```
Comment (Andrewgdewar) [inline src/pages/Landing/index.tsx]: To test this functionality (the error rollback) you should go to apiUtilities.ts and set the chanceOfSuccess to 0.5 or something.

You'll find you have a bug here that duplicates values.

File: src/components/AddCustomerModal.tsx [hunk 4]
```diff
@@ -9,12 +9,55 @@ import {
...
 
 import { SnazzyButton } from './SnazzyButton';
+import { useAppContext } from 'src/middleware';
 
 
 const useStyles = createUseStyles({
   inlineText: {
     display: 'block',
   },
+  darkModeModal: {
```
Comment (Andrewgdewar) [inline src/components/AddCustomerModal.tsx:19]: This is really good, it solves a lot of the problems that currently exist. 🎉

**Comment by Andrewgdewar** [comment]
For Solution 6: 
First, good job targeting the patternfly css variables directly, you've done a great job here. 🚀 

As you clearly have a pretty good understanding of CSS, I'd like to take your learnings to the next level with a small stretch excercise. 

1. First, one problem you haven't identified was that the "ColoredTd.tsx" is applying a color that isn't being applied correctly, please fix that with a "!important" for the imported style.
You should then see the correct colors being applied within the table.

2.  I suggest adding [this](https://chromewebstore.google.com/detail/wcag-color-contrast-check/plnahcmalebffmaghcpcmpaciebdhgdf?hl=en) chrome extension to your browser, and re-updating your darkmode colors (and even the default ones that break contrast rules 🤣 ).

You can also look into WCAG best practices briefly to get an understanding why these rules are followed.

**Comment by Andrewgdewar** [comment]
Overall, really great job!  

⭐ ⭐ ⭐ ⭐ ⭐

File: src/pages/Landing/index.tsx (hunk 1 above)
Comment (dominikvagner) [reply src/pages/Landing/index.tsx]: Good to know! 😀

File: src/pages/Landing/index.tsx (hunk 2 above)
Comment (dominikvagner) [reply src/pages/Landing/index.tsx]: Ooh, that's true, haven't thought of that. 😅

**Comment by dominikvagner** [comment]
> For solution 3: There is no form validation for the submit button. When completing a form field for a modal or really any action that requires inputs from the user, we generally have some logic to prevent the submit button from being used until we have everything we need from them.
> 
> Although not specifically asked in the problem, I would normally expect to see this on any real ticket.

Didn't do that here as it wasn't specifically stated, but that would be a good addition! 😄 In the past I have used the [zod](https://zod.dev/) library for that so that I didn't have to roll my own form validation (and didn't want to add dependecies). What's normally used in our front-end for this? (So that maybe I could try that solution on in this demo repo 🤓)

**Comment by Andrewgdewar** [comment]
> > For solution 3: There is no form validation for the submit button. When completing a form field for a modal or really any action that requires inputs from the user, we generally have some logic to prevent the submit button from being used until we have everything we need from them.
> > Although not specifically asked in the problem, I would normally expect to see this on any real ticket.
> 
> Didn't do that here as it wasn't specifically stated, but that would be a good addition! 😄 In the past I have used the [zod](https://zod.dev/) library for that so that I didn't have to roll my own form validation (and didn't want to add dependecies). What's normally used in our front-end for this? (So that maybe I could try that solution on in this demo repo 🤓)

We use  formik in a few places, but I've used others in the past and am open to trying zod or something new!
Formik can feel a bit clunky at times, and I'm sure there are better/simpler options out there.

**Comment by Andrewgdewar** [comment]
I'll call this one a wrap.

You've completed all tasks as well as the stretch!

I will call out a few of the small #nits I saw though. 
- There is a small bug in your default value of the create modal, preventing deletion of values.
- Although the WCAG colors are now great with dark mode, light mode would need to be tweaked (likely a separate color set for light) to have it fully passing. 

Neither of the above are biggies, but I thought I'd mention them as feedback.

**Review by Andrewgdewar — APPROVED** [review]

**Pull request merged**